internal/mesh/internal/controllers/sidecarproxy/controller.go

```diff
@@ -100,22 +114,41 @@ func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req c
 		return err
 	}
 
-	b := builder.New(req.ID, workloadIdentityRefFromWorkload(workload), trustDomain).
+	// Fetch proxy configuration.
+	proxyCfg, err := dataFetcher.FetchAndMergeProxyConfigurations(ctx, req.ID)
+	if err != nil {
+		rt.Logger.Error("error fetching proxy and merging proxy configurations", "error", err)
+		return err
+	}
+	b := builder.New(req.ID, identityRefFromWorkload(workload), trustDomain, r.dc, proxyCfg).
 		BuildLocalApp(workload.Workload)
 
 	// Get all destinationsData.
-	destinationsRefs := r.cache.DestinationsBySourceProxy(req.ID)
-	destinationsData, statuses, err := dataFetcher.FetchDestinationsData(ctx, destinationsRefs)
+	destinationsRefs := r.destinationsCache.DestinationsBySourceProxy(req.ID)
+	destinationsData, statuses, err := dataFetcher.FetchExplicitDestinationsData(ctx, destinationsRefs)
 	if err != nil {
-		rt.Logger.Error("error fetching destinations for this proxy", "id", req.ID, "error", err)
+		rt.Logger.Error("error fetching explicit destinations for this proxy", "error", err)
 		return err
 	}
 
+	if proxyCfg.GetDynamicConfig() != nil &&
+		proxyCfg.DynamicConfig.Mode == pbmesh.ProxyMode_PROXY_MODE_TRANSPARENT {
+
+		destinationsData, err = dataFetcher.FetchImplicitDestinationsData(ctx, req.ID, destinationsData)
+		if err != nil {
+			rt.Logger.Error("error fetching implicit destinations for this proxy", "error", err)
+			return err
+		}
+	}
```

Comment:
Nit: It feels like we could pull this out into a helper function. Also, I wonder if it would be cleaner to extract the destination merging logic out of `FetchImplicitDestinationsData`.